### PR TITLE
[CenterOfMass] s-quirin redesign

### PR DESCRIPTION
**Pull request by chupins**
- [x] Please check this box if you're not submitting a new macro.   

s-quirin completly redesign the macro (and gives it a lot of improvments)

### CODE REVIEW

**Comment by luzpaz** [comment]
Is there a forum thread associated with the redesign?

**Comment by chupins** [comment]
Yes, it's presented at the end of the original macro thread : https://forum.freecadweb.org/viewtopic.php?p=578665#p578665

File: Information/CenterOfMass.FCMacro [hunk 1]
```diff
@@ -1,891 +1,966 @@
...
-            elif hasattr(sel[s], 'Mesh'):
-                if sel[s].Mesh.Volume == 0:
-                    error_dialog(u'invalid selection (select a solid or a mesh)')
-                    self.destroy()
-                else:
-                    self.solid_count += 1
-                    objs.append(sel[s].Mesh)
+        sel = []
+        objs = self.validSelection()
+        while not self.solid_count and not self.startup_failed:
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:228]: `solid_count` is not a `bool`. Compare with a number to get a `bool`, i.e. `while self.solid_count > 0 and not self.startup_failed`.

File: Information/CenterOfMass.FCMacro [hunk 2]
```diff
@@ -1,891 +1,966 @@
...
+            return self.tree_list.index(s_.Label)
+
+        _sel = FreeCADGui.Selection.getSelection()
+
+        # Add contents of group objects 'groupObjs' (breadth-first search).
+        # For PartDesign objects a Group contains recursive features.
+        # Only nested App:Part are considered at this place.
+        groupObjs = ['App::DocumentObjectGroup', 'App::GeometryPython']
+        i_ = 0
+        while i_ < len(_sel):
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:427]: Can you please have a look at the depth-first search algorithm in `DeepCopy.FCMacro` and see if this is not possible to simplify the code here.

File: Information/CenterOfMass.FCMacro [hunk 3]
```diff
@@ -1,891 +1,966 @@
...
+# - error with draft array of meshes (relevant?)
+# Ideas:
+# - moments of inertia with arrows that allow the user to see the relative magnitudes
 
 import csv
-import glob
 import os
 import math
 
 from PySide import QtCore, QtGui
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:49]: Comment FreeCAD's stealing the PySide module:
`from PySide import QtCore, QtGui  # FreeCAD's PySide!`

File: Information/CenterOfMass.FCMacro [hunk 4]
```diff
@@ -1,891 +1,966 @@
...
+            self.tree_list = [i_.value().text(0) for i_ in list(iterator)]
+            vsel.sort(key=sortSelectionByTree)
+
+        # create valid objects list (e.g. shapes)
+        objs = []
+        for i_ in range(len(vsel)):
+            s_ = vsel[i_]
+            if hasattr(s_, 'Shape'):
+                # local placement correction (transform to global coordinate system)
+                o_ = Part.getShape(s_)    # copy shape for transformation
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:485]: `import Part` is missing.

File: Information/CenterOfMass.FCMacro [hunk 5]
```diff
@@ -1,891 +1,966 @@
...
+        else:
+            tree = FreeCADGui.getMainWindow().findChild(QtGui.QTreeWidget)
+            iterator = QtGui.QTreeWidgetItemIterator(tree, QtGui.QTreeWidgetItemIterator.Editable)
+            # Syntax: [ EXP for x in seq if COND ]
+            self.tree_list = [i_.value().text(0) for i_ in list(iterator)]
+            vsel.sort(key=sortSelectionByTree)
+
+        # create valid objects list (e.g. shapes)
+        objs = []
+        for i_ in range(len(vsel)):
```
Comment (galou) [inline Information/CenterOfMass.FCMacro:481]: Why not `for s_ in vsel:`?

**Review by galou — CHANGES_REQUESTED** [review]
Thanks for the hard work! I didn't test the macro but I guess it works.

Wouldn't it be possible to separate the GUI from the computation of the center of mass?

**Comment by s-quirin** [comment]
Thank you @galou for the in depth analysis, I start looking into it.
Is the "change" of author remarked [in the forum](https://forum.freecadweb.org/viewtopic.php?p=584704#p584704) an issue? It should be consistent to use Github or Forum User name.
@chupins How do you like it?

File: Information/CenterOfMass.FCMacro [hunk 6]
```diff
@@ -1,891 +1,966 @@
...
+        parent.solidLayout.addWidget(self.combo   , sol+1, 2)
+        parent.solidLayout.addWidget(self.spinDens, sol+1, 3)
+        parent.solidLayout.addWidget(self.spinMass, sol+1, 4)
+
+    def init_spinDensity(self, spin, qty, unitForD):
+        spinBoxDigits = 6
+        # don't trigger on_spinDensity_changed
+        # (e.g. when on_comboUnitDensity_changed)
+        spin.blockSignals(True)
+        spin.setMaximum(Units.Quantity('25000 kg/m^3').getValueAs(unitForD))
```
[the remark this answers is not in the document]
Comment (s-quirin) [reply Information/CenterOfMass.FCMacro:178]: It is `MAXIMUM_DENSITY` now

File: Information/CenterOfMass.FCMacro (hunk 5 above)
Comment (s-quirin) [reply Information/CenterOfMass.FCMacro:481]: I need to write back to `vsel` in the last line of the loop

**Comment by galou** [comment]
No problem with the change in author name. Just keep one name per author.

**Comment by chupins** [comment]
Thank a lot for your review Galou.
s-quirin made all your request changes and I just add it to my master

**Comment by galou** [comment]
Thank you both for your contribution and your patience!

File: Information/CenterOfMass.FCMacro (hunk 2 above)
Comment (s-quirin) [reply Information/CenterOfMass.FCMacro:426]: I have several test cases where `DeepCopy.FCMacro` doesn't work as intended:
In the attached file neither for the inner nor the outer Part. There are too many copies.
[CDG_Test_File_schupin.zip](https://github.com/FreeCAD/FreeCAD-macros/files/8440529/CDG_Test_File_schupin.zip)

**Pull request merged**